locator/load_sketch.hh

```diff
@@ -8,87 +8,139 @@
 
 #pragma once
 
+#include "service/tablet_allocator_fwd.hh"
 #include "locator/topology.hh"
 #include "locator/token_metadata.hh"
 #include "locator/tablets.hh"
 #include "utils/stall_free.hh"
 #include "utils/extremum_tracking.hh"
 #include "utils/div_ceil.hh"
+#include "utils/pretty_printers.hh"
 
 #include <absl/container/btree_set.h>
+#include <seastar/util/defer.hh>
 
 #include <optional>
 #include <vector>
 
 namespace locator {
 
+struct disk_usage {
+    using load_type = double; // Disk usage factor (0.0 to 1.0)
+
+    uint64_t capacity = 0;
+    uint64_t used = 0;
+
+    load_type get_load() const {
+        if (capacity == 0) {
+            return 0;
+        }
+        return load_type(used) / capacity;
+    }
+};
+
 /// A data structure which keeps track of load associated with data ownership
 /// on shards of the whole cluster.
 class load_sketch {
     using shard_id = seastar::shard_id;
-    using load_type = ssize_t; // In tablets.
+    using load_type = disk_usage::load_type;
 
     struct shard_load {
         shard_id id;
-        load_type load;
+        disk_usage du;
+        size_t tablet_count = 0;
+
+        // Returns storage utilization for the shard
+        load_type get_load() const {
+            return du.get_load();
+        }
     };
 
     // Less-comparator which orders by load first (ascending), and then by shard id (ascending).
     struct shard_load_cmp {
-        bool operator()(const shard_load& a, const shard_load& b) const {
-            return a.load == b.load ? a.id < b.id : a.load < b.load;
+        bool operator()(const shard_load& shard_a, const shard_load& shard_b) const {
+            auto load_a = shard_a.get_load();
+            auto load_b = shard_b.get_load();
+            return load_a == load_b ? shard_a.id < shard_b.id : load_a < load_b;
         }
     };
 
     struct node_load {
+        std::vector<shard_load> _shards;
         absl::btree_set<shard_load, shard_load_cmp> _shards_by_load;
-        std::vector<load_type> _shards;
-        load_type _load = 0;
+        disk_usage _du;
+        size_t _tablet_count = 0;
 
-        node_load(size_t shard_count) : _shards(shard_count) {
+        // These can be false only when _load_stats != nullptr
+        bool _has_valid_disk_capacity = true;
+        bool _has_all_tablet_sizes = true;
+
+        node_load(size_t shard_count, uint64_t capacity)
+                : _shards(shard_count)
+                , _du({capacity, 0}) {
```

Comment:
nit: please use tagged initialization in case the order of members changes in the future

Reply:
or default initialize and set capacity explicitly same as it is set per shard.